rfcs/testing.md

```diff
@@ -0,0 +1,75 @@
+# RFC: Testing standards
+
+---
+
+_@andrefcdias_
+
+## Summary
+
+The purpose of this RFC is to standardize our testing process and clarify the purpose of each of the testing layers we have. It is supposed to define what is the intended result from tests and why they are required.
+
+## Background
+
+This RFC came from the discussion of unit testing styles in components such as Flex and Text and the debate of having a valuable test vs a "correct" one.
+
+## Problem statement
+
+At the moment, everyone has their own testing standards and there is no common approach to problems like components that only apply styles and need to be tested.
+There is no majority decision on the solution for this problem, being split between Visual Regression tests (VR tests) and Unit Tests (UT).
+
+By defining a common approach, we can make the testing process less ambiguous, as well as provide a guide for the community to follow.
+
+## Detailed Design or Proposal
+
+To start, this is the proposal for defining the goals of each testing layer.
+
+- **Conformance**
+
+  Ensuring the consistency of our components. They test that props like `as`, `ref`, `className` are supported and that the component is rendered.
+
+  These can be achieved simply through the `isConformant` function generated when creating a converged component.
+
+- **Snapshot**
+
+  Stricly DOM structure consistency enforcing. This will guarantee our users don't have breaking changes when they expect a certain structure.
+
+  They are achieved with Jest's `expect(component).toMatchSnapshot()`.
+
+- **Unit**
+
+  Focus on testing the behaviour or functionality. We should use them to guarantee the outcome, wheter this is a styling, behaviour or functionality change.
+  We should try to focus on behaviour, when possible, but mainly on the expected results for the code.
+
+  These are all the additional unit tests defined in your `*.test.ts[x]` file.
+
+- **Integration**
+
+  Share focus with Unit tests, but for cases where you want to test your component/service with other components/services, without mocking those dependencies. Currently we don't differentiate them from Unit tests but a possibility would be to split them into separate files. However, given our tight dependency on other utilities on our components (i.e. `make-styles`) such deliniation might be complicated.
+
+- **Performance**
+
+  Focus on the rendering performance and avoiding it's degradation. They are currently covered by the [flamegrill](https://github.com/microsoft/Flamegrill) utility.
+
+- Bundle size
+
+  Prevent regressions in the bundle size. Currently measured using our custom solution under `/packages/bundle-size`.
+
+- Accessibility
+
+  We do not automate accessibility testing as of the writing of this document.
```

Comment:
I think there is a solution in v0 that we could try to implement or use as a reference to spring up our solution for converged components.

Reply:
We definitely should look at having some static accessibility test wherever possible. Perhaps @smhigley can help define what can be statically tested?

Reply:
we should adopt industry standard tooling 
- `axe` https://github.com/dequelabs/axe-core-npm
- lint rules https://github.com/jsx-eslint/eslint-plugin-jsx-a11y

nice recap https://web.dev/accessibility-auditing-react/

Reply:
I would add that testing almost any keyboard/focus requirements for components can only really be done in the E2E phase inside the browser.

Reply:
I agree with this, accessibility is part of user interaction testing.
Do we have any types of a11y E2E tests to give as an example her?

Reply:
There's also a very old attempt at integrating axe using `axe-puppeteer` in v8 under `apps/a11y-tests`. It ran on all the component examples (not sure if this is the right approach) and saved snapshots of the current errors for each example (again not sure about this approach), but was disabled due to issues with churn in generated classNames and IDs (should have been fixable, but nobody got around to it because IIRC it was basically an FHL project). Not sure how useful this will be going forward but wanted to raise awareness that it exists.

Reply:
> I agree with this, accessibility is part of user interaction testing.
Do we have any types of a11y E2E tests to give as an example here?

`Menu.e2e.ts` - contains a lot of tests that includes keyboarding scenarios